src/detect-modbus.c

```diff
@@ -329,7 +360,12 @@ static DetectModbus *DetectModbusFunctionParse(DetectEngineCtx *de_ctx, const ch
             if (modbus->subfunction == NULL)
                 goto error;
 
-            *(modbus->subfunction) = atoi((const char*) arg);
+            if (StringParseUint16(&(*modbus->subfunction), 10, 0, (const char *)arg) < 0) {
```

Comment:
Suggest making `subfunction` a `uint16_t` instead of a pointer. This would result in no-net increase of space and would allow the checks and free of this to be removed and the call here would simply pass the address of `subfunction` rather than the address of the pointer.

Reply:
I think `subfunction` is a pointer because every modbus message does not have a subfunction, cf file detect-engine-modbus.c
So, you need one more bit in the global structure to say if there is a sub function or not.

I would say this is not the point of this PR

Reply:
It would simplify the code and result in no-net increase in space.

Reply:
I agree that it is good for space.
But how do you @jlucovsky make a simpler code in detect-engine-modbus.c ?

The 2 solutions I see are :
- using a int32_t for `subfunction` and having `-1` value by default to say when there is no subfunction
- add a boolean `has_subfunction ` to the structure to say if the value is a real one

Reply:
I suggest the second option. Avoiding the allocation and memory free  will be possible.

Thus, add a boolean, initialize to `false` and only set when the sub-function is present.

Reply:
Should it be part of this PR ?

Reply:
> Should it be part of this PR ?

No, this should be a separate PR. This PR is about simply converting the string to int conversions.